### PR TITLE
Sapling transaction primitive data.

### PR DESCRIPTION
**Pull request by furszy**
Work coming from #1798. Decoupled commits related to the Sapling transaction primitive data and the new signature hash to make the review a little bit easier there. Plus, squashed some of them.

List of commits taken from #1798:

Primitive data:

* Add serialization for primitive boost::optional<T>. —> 659cc558
* Transaction: Implement BaseOutPoint class.  —> f17b23c7
* Backport zkSNARK compression. —> 5dbc3c9
* Basic Sapling transaction primitive data --> c49594f
* Add serialization for std:list objects. —> 7cc57f7
* sapling_transaction.h migrated to the new serialization —> dbb618f6
* Fix sighash_tests —> e1e33163 
* Transaction GetShieldedValueIn & GetValueOut back port —> 0fbe0236
* BugFix in GetShieldedValueIn —> 57ee1bc6
* Missing SaplingOutPoint::ToString() method implemented —> e1fe5bf7
* Transaction: add sapling data to toString —> 26ce38b5
* make SaplingTxData Optional in transaction primitive  — PARTIALLY  —> c3b1aedc
* Cleaning compiler warnings — PARTIALLY —> 644c381f
* transaction: fix sapling version to 2. —> 8fc595f5
* Transaction: GetDustThreshold method implemented. —> dadf199c.
* Transaction:hasSaplingData check is enough checking for vShieldedOutput and vShieldedSpend emptiness. —> 57d3236f

Signature hash:

* Introduce SignatureHash personalization updates —> 550217f7
* Enable sigversion for Sapling transactions. —> c04abe1d 
* Sapling signature hash (Custom version of ZIP 243) —> eac206e9
* Not try to serialize empty fields in Sapling signature hash  --->  ce5a467

### CODE REVIEW

File: src/script/sign.cpp [hunk 1]
```diff
@@ -26,7 +26,7 @@ bool TransactionSignatureCreator::CreateSig(std::vector<unsigned char>& vchSig,
 
     uint256 hash;
     try {
-        hash = SignatureHash(scriptCode, *txTo, nIn, nHashType, amount, sigversion);
```
Comment (random-zebra) [inline src/script/sign.cpp:29]: This removal makes the  `sigversion` parameter of `CreateSig` unused.
We could remove it from the whole call stack: `Sign1`, `SignN`, and finally `SignStep` (which is also always using  `SIGVERSION_BASE`).

File: src/script/interpreter.cpp [hunk 2]
```diff
@@ -1092,6 +1092,12 @@ const unsigned char PIVX_SEQUENCE_HASH_PERSONALIZATION[crypto_generichash_blake2
         {'P','I','V','X','S','e','q','u','e','n','c','H','a','s','h'};
 const unsigned char PIVX_OUTPUTS_HASH_PERSONALIZATION[crypto_generichash_blake2b_PERSONALBYTES] =
         {'P','I','V','X','O','u','t','p','u','t','s','H','a','s','h'};
+const unsigned char ZCASH_SHIELDED_SPENDS_HASH_PERSONALIZATION[crypto_generichash_blake2b_PERSONALBYTES] =
+        {'P','I','V','X','S','S','p','e','n','d','s','H','a','s','h'};
+const unsigned char ZCASH_SHIELDED_OUTPUTS_HASH_PERSONALIZATION[crypto_generichash_blake2b_PERSONALBYTES] =
+        {'P','I','V','X','S','O','u','t','p','u','t','H','a','s','h'};
```
Comment (random-zebra) [inline src/script/interpreter.cpp:1099]: nit: maybe PIVX-fy these  constants, same as the previous ones.

File: src/script/interpreter.cpp [hunk 3]
```diff
@@ -1179,15 +1219,25 @@ uint256 SignatureHash(const CScript& scriptCode, const CTransaction& txTo, unsig
         // Input prevouts/nSequence (none/all, depending on flags)
         ss << hashPrevouts;
         ss << hashSequence;
+        // Outputs (none/one/all, depending on flags)
+        ss << hashOutputs;
+        if (sigversion == SIGVERSION_SAPLING) {
```
Comment (random-zebra) [inline src/script/interpreter.cpp:1225]: this is always true,  we are already inside `if (sigversion == SIGVERSION_SAPLING) {}`
Maybe we should do
```cpp
if (txTo.hasSaplingData()) {}
```
?

File: src/script/interpreter.cpp [hunk 4]
```diff
@@ -1167,6 +1199,14 @@ uint256 SignatureHash(const CScript& scriptCode, const CTransaction& txTo, unsig
             hashOutputs = ss.GetHash();
         }
 
+        if (!txTo.sapData->vShieldedSpend.empty()) {
```
Comment (random-zebra) [inline src/script/interpreter.cpp:1203]: now we know that `sapData` is always initialized... but  in the future this might be trying to access a `nullopt`

**Review by random-zebra — COMMENTED** [review]
Looking pretty good. Code review ACK, modulo some notes and nits.
There is room for some cleanup due to the removal of the `sigversion` parameter from the `SignatureHash`, but it can be addressed in a followup PR.

File: src/script/interpreter.cpp (hunk 4 above)
Comment (furszy) [reply src/script/interpreter.cpp]: done

File: src/script/interpreter.cpp (hunk 3 above)
Comment (furszy) [reply src/script/interpreter.cpp]: good catch, updated.

File: src/script/interpreter.cpp (hunk 2 above)
Comment (furszy) [reply src/script/interpreter.cpp]: done

File: src/script/sign.cpp (hunk 1 above)
Comment (furszy) [reply src/script/sign.cpp:29]: good eye, done in b4c8c8e

File: src/primitives/transaction.h [hunk 5]
```diff
@@ -257,6 +257,8 @@ class CTransaction
     std::vector<CTxOut> vout;
     const uint32_t nLockTime;
     SaplingTxData sapData;
+    // test-only flag.
+    bool useSapData{true};
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/primitives/transaction.h]: not needed anymore. It's just a remanent of a previous version to the temporarily modified `RandomTransaction`

**Comment by furszy** [comment]
Great feedback!, PR updated.

**Comment by furszy** [comment]
Updated removing the following commit:
`Enable sigversion for Sapling transactions. —> c04abe1`

Instead of defining the SigVersion inside `SignatureHash`, it's better to pass it as an argument, moving the protocol upgrade enforcement check/decision to the caller side.
TODO: This change affects 078f7005 which will need to input `SIGVERSION_SAPLING`.

**Review by random-zebra — COMMENTED** [review]
The global flag to guard v2 txes looks good now.

But have found more instances where we try to access `sapData` without first checking if it's initialized.

In `GetShieldedSpendsHash` and `GetShieldedOutputsHash` we should probably also check first the version of `txTo`, and return early (maybe just `UINT256_ZERO`, instead of the hash of an empty stream) if either `IsSapling()` or `hasSaplingData()` is false.

One more thing.
The hack in the unit test (a3eaeb07cfcc3ebab65e4f3269c5bdb47c7e9028) can be removed now that the version serialization is properly guarded.

**Comment by furszy** [comment]
Done, updated based on the feedback.
`GetShieldedSpendsHash` and `GetShieldedOutputsHash` are both protected on the caller side.

**Comment by random-zebra** [comment]
> `GetShieldedSpendsHash` and `GetShieldedOutputsHash` are both protected on the caller side.

Looking good.
Maybe it's worth to `assert(txTo.sapData)` (or return an error if it's null) inside them,  just to prevent possible future hard-to-discover segfaults if someone introduces code that calls these functions without checking `txTo` first?

**Comment by furszy** [comment]
done, assertion preventing possible future hard-to-discover segfaults implemented.

**Review by random-zebra — APPROVED** [review]
Nice. :beers: ACK 2fff9ec9299ff7d2fd4cc3281c1967b544c031fa

**Review by Fuzzbawls — APPROVED** [review]
utACK 2fff9ec9299ff7d2fd4cc3281c1967b544c031fa

**Pull request merged**